### DISCUSSION TITLE
Is it possible to configure nested services?

### DISCUSSION THREAD

**Discussion by derrickpelletier** [statelyai/xstate#1308] [General] [answered]
Having a hard time finding this in the docs or searching issues/discussions.

I have a repeated substate for some complicated polling behaviour across a few machines, i'd like to define it as a reusable machine/service for each of those primary machines.

If I define a machine, with an invocation to a child machine, if that child machine has a service where `src` is a string, it seems i'm not able to configure that service when i use the parent machine.

```javascript
const complicatedPollMachine = Machine({
  // ...
  states: {
    bar: {
      invoke: {
        src: 'pollingService',
        // ...
      }
    }
  }
});

const myMachine = Machine({
  // ...
  states: {
    foo: {
      invoke: {
        src: complicatedPollMachine,
        // ...
      }
   }
}
```

Then in my component, where i normally define my services:

```javascript
const [state, send] = useMachine(myMachine, {
  services: {
    pollingService: (ctx, event) => { ... },
  }
});

```

However that doesn't work, as it says it can't find the service `pollingService` for `complicatedPollMachine`
Seems if i want to compose something that could be a re-usable child machine, i'm forced to either:

1. define the services for the child machine when I include the child machine in the parent machine definition?
2. use a string for the child machine `src` and include it in the services, or something that feels less than intuitive for a composed state machine...

```javascript
const [ ... ] = useMachine(myMachine, {
  services: {
    pollMachine: complicatedPollMachine.withConfig({
        services: {
          pollingService: (ctx, event) => { ... }
        }
    })
  }
});
```

-------

Hopefully missing something obvious!

**Comment by davidkpiano** [comment, marked as answer]
There's a couple of ways you can think about this. First, if the polling service is relevant to the parent, then it should be the parent's responsibility to have ownership of that service. What that means is that the invoked `complicatedPollMachine` shouldn't know about the `pollingService`, and only send/receive events from the parent. The parent would then react to received events and direct them to the `pollingService` (that it directly owns) as necessary.

Think about it like a 3rd-party API. You can't (and shouldn't be able to) directly provision that API from whatever invokes it, although you can do things like pass configuration data to it:

```js
invoke: {
  src: complicatedPollMachine,
  data: {
    url: 'https://example.com/pollingapi',
    interval: 2000
  }
}
```

IMO, that would be the best way to go, especially since it reflects programming patterns and best practices that you're probably used to already. And you avoid "affecting" the child machine; that should be a code smell, since it is dealing with implementation details of an invoked machine.

-----

Another alternative would be to get creative with JS and do something like this, if you wanted:

```js
const createMyMachine = (pollingService) => createMachine({
  // ...
  invoke: {
    src: () => complicatedMachine.withConfig({ services: { pollingService } })
  }
});
```

**Reply by derrickpelletier** [reply]
Yeah the second example is definitely what i'm trying to _avoid_ here. But maybe i'm placing too much value in the `src`-as-string feature for testability. I haven't worked with xstate/statecharts enough to know for sure 😀

Your first example definitely makes the "complicatedPollMachine" locked into a specific implementation which I guess I was trying to avoid... but I like your points, going to give that try instead of whatever pattern I was trying to scheme up 

Thanks for your help 🙏

**Reply by thebookins** [reply]
@derrickpelletier I'd love to know if you ever solved this issue? I have a very similar architecture (parent machine that invokes a child machine, where the child machine needs to be passed a service from the React component that instantiates the parent machine). In my case rather than a `pollingService` I need to pass a `playSound` service that uses the [React `useSound` hook](https://www.npmjs.com/package/use-sound).

I have found your sample code from above the only reliable way of getting this to work:
```js
  const [play] = useSound(audio, { sprite, interrupt: true });

  // wrapping play from useSound in a promise that resolves when the audio has finished
  const playSound = (context, event) => new Promise((resolve) => {
    const id = context.sound;
    const duration = sprite[id][1];
    setTimeout(resolve, duration);
    play({ id });
  });

  const [ ... ] = useMachine(myMachine, {
    services: {
      myChildMachine: myChildMachine.withConfig({
        services: { playSound }
      })
    }
  });
```

Are there any issues nesting services in this manner? I also tried the second example from @davidkpiano (passing `playSound` in a `createMyMachine` function) but could not get this to work (I'm assuming because `useSound` takes some time to be fully initialized).
